README.md
# varband

The `varband` package contains the implementations of the variable banding method for learning local dependence and 
estimating large sparse precision matrix in the setting where variables have a natural ordering. 
The details of the method can be found in 
[Yu, Bien (2017) *Learning Local Dependence in Ordered Data* (Journal of Machine Learning Research (2017), 18(42), 1-60)](http://arxiv.org/abs/1604.07451).

`varband` is now on CRAN. To install `varband`, type in R console
```R
install.packages("varband")
```

To install `varband` from [github](http://github.com), which is usually more up-to-date, type in R console
```R
devtools::install_github("hugogogo/varband", build_vignettes = TRUE)
```
Note that the installation above requires using R package [devtools](https://cran.r-project.org/web/packages/devtools/index.html)
(which can be installed using `install.packages("devtools")`).

An example of how to use the package to solve a simple problem is given in the vignette. To see the vignette, after installation,
type in R console
```R
vignette("varband-vignette")
```


NEWS.md
# `varband` 0.9.0 ready for CRAN submission!
# `varband` 0.9.1 allows users to specify a maximum bandwidth in the resulting estimate to fasten the computation.


R/utils.R
#### This file contains all utility functions
#### Not necessarily useful in the package

#' Plot the sparsity pattern of a square matrix
#'
#' Black, white and gray stand for positive, zero and negative respectively
#'
#' @param Mat A matrix to plot.
#' @param main A plot title.
#'
#' @examples
#' set.seed(123)
#' p <- 50
#' n <- 50
#' phi <- 0.4
#' true <- varband_gen(p = p, block = 5)
#' matimage(true)
#' @export
matimage <- function(Mat, main = NULL){
  tmppar <- par(pty = "s")
  image(sign(t(apply(Mat, 2, rev))), axes = FALSE, col = c("gray50","white","black"), main = main)
  par(tmppar)
}

KL <- function(Omega_est, Omega_true){
  # Kullback-Leibler divergence for gaussian likelihood
  # see (14) in Nested Lasso (Levina et,al 2008)
  res <- determinant(x = Omega_est, logarithm = TRUE)$modulus
  res <- -res + determinant(x = Omega_true, logarithm = TRUE)$modulus
  res <- sum(diag(solve(Omega_true, Omega_est))) + res - nrow(Omega_est)
  return(as.numeric(res))
}

Irrep <- function(L){
  # Calculate the irrepresentability condition of true L
  p <- ncol(L)
  Sigma <- solve(crossprod(L))
  res <- c()
  for (r in seq(2, p)){
    # For each row
    Jr <- 0
    for (j in seq(r-1)){
      if(L[r, j] != 0){
        Jr <- j-1
        break
      }
    }
    # Got Ir
    if(Jr > 0){
      Ir <- ((Jr+1):(r-1))
      tmp <- rep(NA, Jr)
      for (ll in seq(Jr)){
        tmp[ll] <- norm(as.matrix(Sigma[ll, Ir]%*%solve(Sigma[Ir, Ir])), "1")
      }
      cat("r = ", r, "  ", "Jr = ", length(tmp), "  ", "Kr = ", length(Ir), fill = TRUE)
      res <- c(res, max(tmp))
    }
  }
  return(max(res))
}

#################################################################
####  The following functions are the first R version of main
####  functions used in the package
####  They have been replaced by their c++ version
#################################################################
# #' Evaluate the proximal operator of the hierarchical group lasso with general weights
# #'
# #' This function solves (7) in the paper for general weight w
# #' by solving its dual by performing Newton's method on at most
# #' r-1 univariate functions.
# #' See Algorithm 1 and Theorem 1 in the online supplemenatry.
# #'
# #' @param y An r-dimensional vector.
# #' @param tau lambda/rho
elliproj_w_R <- function(y, tau){
  # This function performs the ellipsoid projection
  # See supplementary material
  r <- length(y)

  nu <- rep(NA, r-1)
  # pp is the z vector in the paper
  pp <- y
  for(l in seq(r-1)){
    # ww[m] = w_{lm}
    ww <- 1 / (seq(l,1)^2)
    # check if it lies in the ellipsoid
    if (sum((pp[1:l]/ww)^2) <= tau^2){
      nu[l] <- 0
      pp[1:l] <- 0
    }
    else{
      # project onto the elliposid
      f <- function(nu)
        1-tau/sqrt(sum((pp[1:l]/(ww+nu/ww))^2))
      # lower and upper bound of root
      nu.u <- sqrt(sum((ww*pp[1:l])^2))/tau
      nu.l <- max(nu.u-ww[l]^2, 0)
      # find root
      if(abs(f(nu.u)) < SMALL)
        nu[l] <- nu.u
      else if (abs(f(nu.l)) < SMALL)
        nu[l] <- nu.l
      else
        nu[l] <- uniroot(f = f, interval = c(nu.l, nu.u), tol = 1e-15)$root
      pp[1:l] <- pp[1:l] * nu[l] / ( ww^2 + nu[l] )
    }
  }
  return(pp)
}

# #' Evaluate the proximal operator of the hierarchical group lasso with simple weights
# #'
# #' This function solves (7) in the paper for unweighted version(w = 1)
# #' by solving its dual by performing Newton's method on at most
# #' r-1 univariate functions.
# #' See Algorithm 2 in the paper
# #'
# #' @param y An r-dimensional vector.
# #' @param tau lambda/rho
elliproj_u_R <- function(y, tau){
  # This function performs the ellipsoid projection
  # of the unweighted estimator, which is very easy
  # See algorithm 2 in the paper
  r <- length(y)
  # pp is the z vector in the paper
  pp <- y
  for(l in seq(r-1)){
    tmpnorm <- sqrt(sum(pp[1:l]^2))
    if(tmpnorm <= tau)
      pp[1:l] <- 0
    else
      pp[1:l] <- (1-tau/tmpnorm)*pp[1:l]
  }
  return(pp)
}

SMALL <- 1e-15
# #' Compute the varband estimate for a fixed tuning parameter value.
# #'
# #' Solves the main optimization problem in Yu & Bien (2016):
# #' \deqn{min_L -2 \sum_{r=1}^p L_{rr} + tr(SLL^T) + lam * \sum_{r=2}^p P_r(L_{r.})}{min_L -2 sum_{r=1}^p L_{rr} + tr(SLL^T) + lam * sum_{r=2}^p P_r(L_{r.})}
# #' where \deqn{P_r(L_{r.}) = \sum_{\ell = 2}^{r-1} \left(\sum_{m=1}^\ell w_{\ell m}^2 L_{rm}^2\right)^{1/2}}{P_r(L_r.) = sum_{l=2}^{r-1} (sum_m=1^l w^2_lm L^2_rm)^{1/2}}
# #'
# #' The function decomposes into p independent row problems,
# #' each of which is solved by an ADMM algorithm.
# #' see paper for more explanation.
# #' @param S The sample covariance matrix
# #' @param lambda Non-negative tuning parameter. Controls sparsity level.
# #' @param w Logical. Should we use weighted version of the penalty or not? If \code{TRUE}, we use general weight. If \code{FALSE}, use unweighted penalty. Default is \code{FALSE}.
# #' @param init Initial estimate of L. Default is a closed-form diagonal estimate of L.
# #' @return Returns the variable banding estimate of L, where L^TL = Omega.
# #'
# #' @seealso \code{\link{varband_path}} \code{\link{varband_cv}}
# #'
varband_R <- function(S, lambda, init = NULL, w = FALSE){
  p <- ncol(S)
  # check that S is square
  stopifnot(p == nrow(S))
  if (is.null(init))
    init <- diag(1/sqrt(diag(S)))

  L <- matrix(0,p,p)
  L[1,1] <- 1/(sqrt(S[1,1]))

  # for the second to the p-th row
  for(r in seq(2,p)){
    L[r,] <- c(rowadmm_R(S = S[1:r,1:r],
                         init_row = init[r,1:r],
                         lambda = lambda, w = w),
               rep(0,p-r))
  }
  return(L)
}

# #' Compute one row of varband estimate for a fixed tuning parameter
# #'
# #'This function solve the following r-th row estimation problem \deqn{min_{beta_r>0} -2 log beta_r + 1/n ||X beta||^2 + lambda P(beta)}
# #' using an ADMM algorithm with changing rho.
# #'
# #' See algorithm 1 in the paper.
# #'
# #' @param S An r-by-r submatrix of sample covariance matrix.
# #' @param init_row The initial estimate of the row.
# #' @param lambda Non-negative tuning parameter. Controls sparsity level.
# #' @param w Logical. Should we use weighted version of the penalty or not? If \code{TRUE}, we use general weight. If \code{FALSE}, use unweighted penalty. Default is \code{FALSE}.
# #' @param tol Tolerance for convergence.
# #' @param itermax Maximum number of iterations of ADMM to perform.
rowadmm_R <- function(S, init_row, lambda, w = FALSE, tol = 1e-4, itermax = 1e5){
  # This function solve the following row estimation problem
  # \min_{\beta_r>0} -2 log \beta_r + 1/n ||X\beta||^2
  # + \lambda P(\beta)
  # using an ADMM algorithm with changing rho

  r <- ncol(S)
  stopifnot(r == nrow(S))
  # could use a lower tolerance for unweighted version
  if (!w)
    tol <- 1e-8

  # Default parameter in ADMM
  tolabs <- tol
  tolrel <- tol
  # Changing rho
  rho <- 2
  mu <- 10
  inc <- 2
  dec <- 2

  # Initialize the result
  beta <- init_row
  gamma <- init_row

  # dual variable
  u <- rep(0,r)

  S_inv <- inverse_update_R(S = as.matrix(S[-r,-r]),
                            r = r, rho = rho)
  for (i in seq(itermax)) {
    # Primal&Dual Updates
    beta_new <- close_update_R(S = S, S_inv = S_inv,
                               r = r, rho = rho, u = u,
                               gamma = gamma)
    if (w)
      gamma_new <- elliproj_w_R(y = beta_new + u/rho,
                                tau = lambda/rho)
    else
      gamma_new <- elliproj_u_R(y = beta_new + u/rho,
                                tau = lambda/rho)

    u <- u + rho*(beta_new - gamma_new)

    # check convergence See pp 22 Boyd(2011)
    # primal residual
    pres <- sqrt(crossprod(beta_new - gamma_new))
    # dual residual
    dres <- rho*sqrt(crossprod(gamma_new - gamma))
    # primal tolerance
    peps <- tolabs*sqrt(r) +
      tolrel*max(sqrt(crossprod(beta_new)),
                 sqrt(crossprod(gamma_new)))
    # dual tolerance
    deps <- tolabs*sqrt(r) + tolrel*sqrt(crossprod(u))

    if(pres <= peps & dres <= deps)
    {
      #cat("ADMM converges after",i,"iterations",fill=TRUE)
      break
    }
    else{
      # if not, update estimates and rho
      beta <- beta_new
      gamma <- gamma_new

      # Update rho if needed and corresponding S_inv
      if(pres > mu*dres){
        rho <- rho*inc
        S_inv <- inverse_update_R(S = as.matrix(S[-r,-r]),
                                  r = r, rho = rho)
      }
      else if(dres > mu*pres){
        rho <- rho/dec
        S_inv <- inverse_update_R(S = as.matrix(S[-r,-r]),
                                  r = r, rho = rho)
      }
    }
  }
  if(i==itermax)
    cat("ADMM fails to converge",fill=TRUE)
  return (gamma_new)
}

inverse_update_R <- function(S, r, rho){
  if(r==1){
    S_inv <- as.matrix(1/(2*S[1,1] + rho))
    return(S_inv)
  }
  else{
    S_inv <- 2*S
    diag(S_inv) <- diag(S_inv) + rho
    S_inv <- solve(S_inv)
    return(S_inv)
  }
}

# #' Close-form update of beta in Algorithm 1
# #'
# #' This function solves (6) in the paper with a closed form solution
# #'
# #' @param S An r-by-r submatrix of sample covariance matrix.
# #' @param S_inv inverse of (2S_{-r,-r} + rho I)
# #' @param r row index
# #' @param rho parameter rho used in ADMM
# #' @param u dual variable in ADMM
# #' @param gamma priaml variable in ADMM
close_update_R <- function(S, S_inv, r, rho, u, gamma){
  # This performs the closed updated for beta
  # see Section 3 in the paper
  vec.tmp <- S_inv%*%S[-r,r]
  A <- 4*crossprod(vec.tmp,S[-r,r]) - 2*S[r,r] - rho
  B <- 2*crossprod(vec.tmp, (u[-r] - rho*gamma[-r])) - u[r] + rho*gamma[r]
  res <- rep(0,r)
  # beta_r
  res[r] <- (-sqrt(B^2 - 8 * A) - B)/(2 * A)
  # beta_[-r]
  res[-r] <- -2*res[r]*vec.tmp - S_inv%*%(u[-r] - rho*gamma[-r])
  return(res)
}



R/model_gen.R
#' Generate random samples.
#'
#' Generate \code{n} random samples from multivariate Gaussian distribution N(0, (L^TL)^{-1})
#'
#' @param L p-dimensional inverse Cholesky factor of true covariance matrix.
#' @param n number of samples to generate.
#' @return returns a n-by-p matrix with each row a random sample generated.
#' @examples
#' set.seed(123)
#' true <- varband_gen(p = 50, block = 5)
#' x <- sample_gen(L = true, n = 100)
#' @export
sample_gen<-function(L, n){
  #### This function generates normal random vectors
  # Y ~ N(0, (L^T L)^{-1}) = N(0, L^{-1}L^{-T})
  # Sigma = L^{-1}L^{-T}
  p <- nrow(L)
  # Z~N(0,1_p)
  # Z<-mvrnorm(n, rep(0,p), diag(rep(1,p)))
  Z <- matrix(rnorm(n * p), n, p)
  # X[i,]%*%L^{-T} ~ N (0, L^{-1}L^{-T}) = N(0, Sigma)
  X <- t(solve(L, t(Z)))
  return(X)
}

#' Generate an autoregressive model.
#'
#' Generate lower triangular matrix with strict bandwidth. See, e.g., Model 1 in the paper.
#'
#' @param p the dimension of L
#' @param phi_vec a K-dimensional vector for off-diagonal values
#' @return a p-by-p strictly banded lower triangular matrix
#' @examples
#' true_ar <- ar_gen(p = 50, phi = c(0.5, -0.4, 0.1))
#' @export
ar_gen<-function(p, phi_vec){
  set.seed(123)
  ii <- toeplitz(1:p)
  K <- length(phi_vec) + 1
  L <- ii <= K

  for (k in seq(2, K)) L[ii == k] <- phi_vec[k - 1]
  diag(L) <- rep(0, p)
  L <- L - upper.tri(L) * L

  L <- diag(rep(1, p)) - L
  # Note that Omega = L^T L
  # NEWLY ADDED
  L <- diag(1/runif(p, 2, 5)) %*% L
  return(L)
}

#' Generate a model with variable bandwidth.
#'
#' Generate lower triangular matrix with variable bandwidth. See, e.g., Model 2 and 3 in the paper.
#' @param p the dimension of L
#' @param block the number of block diagonal structures in the resulting model, assumed to divide p
#' @return a p-by-p lower triangular matrix with variable bandwidth
#' @examples
#' set.seed(123)
#' # small block size (big number of blocks)
#' true_small <- varband_gen(p = 50, block = 10)
#' # large block size (small number of blocks)
#' true_large <- varband_gen(p = 50, block = 2)
#' @export
varband_gen <- function(p, block = 10){
  set.seed(123)
  L <- matrix(0, p, p)
  stopifnot(p%%block == 0)
  block_size <- p / block
  for(k in seq(block)){
    L[((k-1) * block_size + 1):(k * block_size),
      ((k-1)*block_size + 1):(k * block_size)] <- block_gen(block_size)
  }
  diag(L) <- 1
  L <- diag(1/runif(p, 2, 5)) %*% L
  return(L)
}

block_gen <- function(block_size){
  smallL <- matrix(0, block_size, block_size)
  for(i in seq(2, block_size)){
    flag <- rbinom(1, 1, 0.5)
    if (flag == 1){
      Ji <- sample(i - 1, 1)
      sign <- ((runif(i - 1, 0, 1) > 0.5) - 1/2 ) * 2
      smallL[i, 1 : (i - 1)] <- sign * runif(i - 1, 0.1, 0.4)
      smallL[i, 1 : Ji] <- 0
    }
    else
      smallL[i, 1:(i - 1)] <- 0
  }
  return(smallL)
}

#' Generate a model with block-diagonal structure
#'
#' @param p the dimension of L
#' @return a p-by-p lower triangular matrix with block-diagonal structure from p/4-th row to 3p/4-th row
#' @examples
#' set.seed(123)
#' true_L_block_diag <- block_diag_gen(p = 50)
#' @export
block_diag_gen <- function(p){
  L <- matrix(0, p, p)
  set.seed(123)
  for(i in seq(p / 4 + 2, p * 3 / 4)){
    sign <- ((runif(i - p / 4 - 1, 0, 1) > 0.5) - 1/2 ) * 2
    L[i, (p / 4 + 1) : (i - 1)] <- sign * runif(i - p / 4 - 1, 0.1, 0.2)
  }
  diag(L) <- 1
  L <- diag(1/runif(p, 2, 5)) %*% L
  return(L)
}


R/RcppExports.R
# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

soft_threshold <- function(beta, lam, r, result) {
    invisible(.Call('varband_soft_threshold', PACKAGE = 'varband', beta, lam, r, result))
}

close_update <- function(S, S_inv, r, rho, u, gamma, res) {
    invisible(.Call('varband_close_update', PACKAGE = 'varband', S, S_inv, r, rho, u, gamma, res))
}

inverse_update <- function(S, rho, S_inv) {
    invisible(.Call('varband_inverse_update', PACKAGE = 'varband', S, rho, S_inv))
}

elliproj_u <- function(y, tau, pp) {
    invisible(.Call('varband_elliproj_u', PACKAGE = 'varband', y, tau, pp))
}

rootfind <- function(pp, ww, tau, l) {
    .Call('varband_rootfind', PACKAGE = 'varband', pp, ww, tau, l)
}

elliproj_w <- function(y, tau, pp) {
    invisible(.Call('varband_elliproj_w', PACKAGE = 'varband', y, tau, pp))
}

rowadmm <- function(S, init_row, lambda, w = FALSE, tol = 1.0e-4, itermax = 1e+6L) {
    .Call('varband_rowadmm', PACKAGE = 'varband', S, init_row, lambda, w, tol, itermax)
}

rowadmm_lasso <- function(S, init_row, lambda, tol = 1.0e-4, itermax = 1e+6L) {
    .Call('varband_rowadmm_lasso', PACKAGE = 'varband', S, init_row, lambda, tol, itermax)
}

#' Compute the varband estimate for a fixed tuning parameter value with different penalty options.
#'
#' Solves the main optimization problem in Yu & Bien (2017):
#' \deqn{min_L -2 \sum_{r=1}^p L_{rr} + tr(SLL^T) + lam * \sum_{r=2}^p P_r(L_{r.})}{min_L -2 sum_{r=1}^p L_{rr} + tr(SLL^T) + lam * sum_{r=2}^p P_r(L_{r.})}
#' where \deqn{P_r(L_{r.}) = \sum_{\ell = 2}^{r-1} \left(\sum_{m=1}^\ell w_{\ell m}^2 L_{rm}^2\right)^{1/2}}{P_r(L_r.) = sum_{l=2}^{r-1} (sum_m=1^l w^2_lm L^2_rm)^{1/2}}
#' or \deqn{P_r(L_{r.}) = \sum_{\ell = 1}^{r-1} |L_{r\ell}|}
#'
#' The function decomposes into p independent row problems,
#' each of which is solved by an ADMM algorithm.
#' see paper for more explanation.
#' @param S The sample covariance matrix
#' @param lambda Non-negative tuning parameter. Controls sparsity level.
#' @param init Initial estimate of L. Default is a closed-form diagonal estimate of L.
#' @param K Integer between 0 and p - 1 (default), indicating the maximum bandwidth in the resulting estimate. A small value of K will result in a sparse estimate and small computing time.
#' @param w Logical. Should we use weighted version of the penalty or not? If \code{TRUE}, we use general weight. If \code{FALSE}, use unweighted penalty. Default is \code{FALSE}.
#' @param lasso Logical. Should we use l1 penalty instead of hierarchical group lasso penalty? Note that by using l1 penalty, we lose the banded structure in the resulting estimate. Default is \code{FALSE}.
#' @return Returns the variable banding estimate of L, at most K banded, where L^TL = Omega.
#'
#' @examples
#' set.seed(123)
#' n <- 50
#' true <- varband_gen(p = 50, block = 5)
#' x <- sample_gen(L = true, n = n)
#' S <- crossprod(scale(x, center = TRUE, scale = FALSE)) / n
#' init <- diag(1/sqrt(diag(S)))
#' # unweighted estimate
#' L_unweighted <- varband(S, lambda = 0.1, init, w = FALSE)
#' # at most 10-banded unweighted estimate
#' L_unweighted_K10 <- varband(S, lambda = 0.1, init, w = FALSE, K = 10)
#' # weighted estimate
#' L_weighted <- varband(S, lambda = 0.1, init, w = TRUE)
#' # lasso estimate
#' L_lasso <- varband(S, lambda = 0.1, init, w = TRUE, lasso = TRUE)
#' @seealso \code{\link{varband_path}} \code{\link{varband_cv}}
#'
#' @export
varband <- function(S, lambda, init, K = -1L, w = FALSE, lasso = FALSE) {
    .Call('varband_varband', PACKAGE = 'varband', S, lambda, init, K, w, lasso)
}



R/varband_path.R
#' Solve main optimization problem along a path of lambda
#'
#' Compute the varband estimates along a path of tuning parameter values.
#'
#' @param S The sample covariance matrix
#' @param w Logical. Should we use weighted version of the penalty or not? If \code{TRUE}, we use general weight. If \code{FALSE}, use unweighted penalty. Default is \code{FALSE}.
#' @param lasso Logical. Should we use l1 penalty instead of hierarchical group lasso penalty? Note that by using l1 penalty, we lose the banded structure in the resulting estimate. And when using l1 penalty, the becomes CSCS (Convex Sparse Cholesky Selection) introduced in Khare et al. (2016). Default value for \code{lasso} is \code{FALSE}.
#' @param K Integer between 0 and p - 1 (default), indicating the maximum bandwidth in the resulting estimate. A small value of K will result in a sparse estimate and small computing time.
#' @param lamlist A list of non-negative tuning parameters \code{lambda}.
#' @param nlam If lamlist is not provided, create a lamlist with length \code{node}. Default is 60.
#' @param flmin if lamlist is not provided, create a lamlist with ratio of the smallest and largest lambda in the list. Default is 0.01.
#' @return A list object containing \describe{
#' \item{path: }{A array of dim (\code{p}, \code{p}, \code{nlam}) of estimates of L}
#' \item{lamlist: }{a grid values of tuning parameters}
#' }
#' @examples
#' set.seed(123)
#' n <- 50
#' true <- varband_gen(p = 50, block = 5)
#' x <- sample_gen(L = true, n = n)
#' S <- crossprod(scale(x, center = TRUE, scale = FALSE))/n
#' path_res <- varband_path(S = S, w = FALSE, nlam = 40, flmin = 0.03)
#' @export
#'
#' @seealso \code{\link{varband}} \code{\link{varband_cv}}
varband_path <- function(S, w = FALSE, lasso = FALSE, K = -1,
                         lamlist = NULL, nlam = 60, flmin = 0.01){
  p <- ncol(S)
  stopifnot(p == nrow(S))

  if (is.null(lamlist)) {
    lam_max <- lammax(S = S)
    lamlist <- pathGen(nlam = nlam, lam_max = lam_max,
                       flmin = flmin, S = S)
  } else {
    nlam <- length(lamlist)
  }
  result<- array(NA, c(p, p, nlam))

  for (i in seq(nlam)) {
    if(i==1){
      #      cat(i)
      result[, , i] <- diag(1/sqrt(diag(S)))
    }
    else
    {
      #      cat(i)
      result[, , i] <- varband(S = S, lambda = lamlist[i],
                               init = result[, , i-1],
                               K = K, w = w, lasso = lasso)
    }
  }
  return(list(path = result, lamlist = lamlist))
}

lammax <- function(S){
  #### This function calculates the max value in the tuning parameter list
  # such that the estimator L_{\lambda} is a diagonal matrix
  # NOTE: this is not necessarily true, but generally
  # a upper bound of the value we are looking for.

  # Args:
  #     S: the p-by-p sample covariance matrix

  p <- ncol(S)
  sighat <- rep(NA, p-1)
  for (r in seq(2, p)){
    sighat[r-1] <- max(abs(S[(1:(r-1)), r]))/sqrt(S[r, r])
  }
  2 * max(sighat)
}

pathGen <- function(nlam, lam_max, flmin, S){
  # Generate a path of lambda, with
  # nlam/2 decreasing exponentially
  # nlam/2 decreasing linearly
  # lam_max <- lammax(S)
  lamlist_lin <- lam_max * exp(seq(0, log(flmin), length = nlam/2))
  lamlist_exp <- seq(lam_max - 1e-8, lam_max*flmin - 1e-8, length.out = nlam/2)
  return(sort(unique(c(lamlist_lin, lamlist_exp)), decreasing = T))
}


R/varband_cv.R
#' Perform nfolds-cross validation
#'
#' Select tuning parameter by cross validation according to the likelihood on testing data, with and without refitting.
#'
#' @param x A n-by-p sample matrix, each row is an observation of the p-dim random vector.
#' @param w Logical. Should we use weighted version of the penalty or not? If \code{TRUE}, we use general weight. If \code{FALSE}, use unweighted penalty. Default is \code{FALSE}.
#' @param lasso Logical. Should we use l1 penalty instead of hierarchical group lasso penalty? Note that by using l1 penalty, we lose the banded structure in the resulting estimate. And when using l1 penalty, the becomes CSCS (Convex Sparse Cholesky Selection) introduced in Khare et al. (2016). Default value for \code{lasso} is \code{FALSE}.
#' @param K Integer between 0 and p - 1 (default), indicating the maximum bandwidth in the resulting estimate. A small value of K will result in a sparse estimate and small computing time.
#' @param lamlist A list of non-negative tuning parameters \code{lambda}.
#' @param nlam If lamlist is not provided, create a lamlist with length \code{nulam}. Default is 60.
#' @param flmin If lamlist is not provided, create a lamlist with ratio of the smallest and largest lambda in the list equal to \code{flmin}. Default is 0.01.
#' @param folds Folds used in cross-validation
#' @param nfolds If folds are not provided, create folds of size \code{nfolds}.
#'
#' @return A list object containing \describe{
#' \item{errs_fit: }{A \code{nlam}-by-\code{nfolds} matrix of negative Gaussian log-likelihood values on the CV test data sets. \code{errs[i,j]} is negative Gaussian log-likelihood values incurred in using \code{lamlist[i]} on fold \code{j}}.
#' \item{errs_refit: }{A \code{nlam}-by-\code{nfolds} matrix of negative Gaussian log-likelihood values of the refitting.}
#' \item{folds: }{Folds used in cross validation.}
#' \item{lamlist: }{\code{lambda} grid used in cross validation.}
#' \item{ibest_fit: }{index of \code{lamlist} minimizing CV negative Gaussian log-likelihood.}
#' \item{ibest_refit: }{index of \code{lamlist} minimizing refitting CV negative Gaussian log-likelihood.}
#' \item{i1se_fit: }{Selected value of \code{lambda} using the one-standard-error rule.}
#' \item{i1se_refit: }{Selected value of \code{lambda} of the refitting process using the one-standard-error rule.}
#' \item{L_fit: }{Estimate of L corresponding to \code{ibest_fit}.}
#' \item{L_refit: }{Refitted estimate of L corresponding to \code{ibest_refit}.}
#' }
#'
#' @examples
#' set.seed(123)
#' p <- 50
#' n <- 50
#' true <- varband_gen(p = p, block = 5)
#' x <- sample_gen(L = true, n = n)
#' res_cv <- varband_cv(x = x, w = FALSE, nlam = 40, flmin = 0.03)
#' @export
#'
#' @seealso \code{\link{varband}} \code{\link{varband_path}}
varband_cv <- function(x, w = FALSE, lasso = FALSE, K = -1,
                       lamlist = NULL, nlam = 60, flmin = 1e-2,
                       folds = NULL, nfolds = 5) {
  n <- nrow(x)
  p <- ncol(x)

  S <- crossprod(scale(x, center=TRUE, scale=FALSE)) / n
  if(is.null(folds))
    folds <- makefolds(n, nfolds = nfolds)
  nfolds <- length(folds)

  if (is.null(lamlist)) {
    lam_max <- lammax(S = S)
    lamlist <- pathGen(nlam = nlam, lam_max = lam_max,
                       flmin = flmin, S = S)
  } else {
    nlam <- length(lamlist)
  }

  errs_fit <- matrix(NA, nlam, nfolds)
  errs_refit <- matrix(NA, nlam, nfolds)

  # error function is the negative log Gaussian likelihood
  for (i in seq(nfolds)) {
    # train on all but i-th fold:
    x_tr <- x[-folds[[i]],]
    meanx <- colMeans(x_tr)
    x_tr <- scale(x_tr, center = meanx, scale = FALSE)
    S_tr <- crossprod(x_tr) / (dim(x_tr)[1])

    path_fit <- varband_path(S = S_tr, w = w, lasso = lasso, K = K,
                             lamlist = lamlist)$path
    path_refit <- refit_path(S = S_tr, path = path_fit)

    # evaluate this on left-out fold:
    x_te <- x[folds[[i]], ]
    x_te <- scale(x_te, center = meanx, scale = FALSE)
    S_te <- crossprod(x_te) / (dim(x_te)[1])

    for (j in seq(nlam)) {
      errs_fit[j, i] <- likelihood(crossprod(path_fit[, , j]), S_te)
      errs_refit[j, i] <- likelihood(crossprod(path_refit[, , j]), S_te)
    }
  }

  m_fit <- rowMeans(errs_fit)
  se_fit <- apply(errs_fit, 1, sd) / sqrt(nfolds)
  m_refit <- rowMeans(errs_refit)
  se_refit <- apply(errs_refit, 1, sd) / sqrt(nfolds)
  ibest_fit <- which.min(m_fit)
  i1se_fit <- min(which(m_fit < m_fit[ibest_fit] + se_fit[ibest_fit]))
  ibest_refit <- which.min(m_refit)
  i1se_refit <- min(which(m_refit < m_refit[ibest_refit] + se_refit[ibest_refit]))


  fit_cv <- varband(S = S, lambda = lamlist[ibest_fit],
                    init = path_fit[, , ibest_fit], K = K,
                    w = w, lasso = lasso)
  refit_cv <- varband(S = S, lambda = lamlist[ibest_refit],
                      init = path_refit[, , ibest_refit], K = K,
                      w = w, lasso = lasso)
  refit_cv <- refit_matrix(S = S, mat = refit_cv)

  return(list(errs_fit = errs_fit, errs_refit = errs_refit,
              folds = folds, lamlist = lamlist,
              ibest_fit = ibest_fit, ibest_refit = ibest_refit,
              i1se_fit = i1se_fit, i1se_refit = i1se_refit,
              L_fit = fit_cv, L_refit = refit_cv))
}

makefolds <- function(n, nfolds) {
  # divides the indices 1:n into nfolds random folds of about the same size.
  nn <- round(n / nfolds)
  sizes <- rep(nn, nfolds)
  sizes[nfolds] <- sizes[nfolds] + n - nn * nfolds
  b <- c(0, cumsum(sizes))
  ii <- sample(n)
  folds <- list()
  for (i in seq(nfolds))
    folds[[i]] <- ii[seq(b[i] + 1, b[i + 1])]
  folds
}

likelihood <- function (Omega, S){
  # Calculate the negative log-Gaussian likelihood with
  # precision matrix Omega and sample covariance S
  return(-determinant(Omega, logarithm = TRUE)$modulus[1] + sum(S*Omega))
}
